File: server/module/module-erp/erp-server/src/service/erp_product_unit.rs
use sea_orm::{DatabaseConnection, EntityTrait, ColumnTrait, ActiveModelTrait, PaginatorTrait, QueryOrder, QueryFilter, Condition};
use crate::model::erp_product_unit::{Model as ErpProductUnitModel, ActiveModel as ErpProductUnitActiveModel, Entity as ErpProductUnitEntity, Column};
use erp_model::request::erp_product_unit::{CreateErpProductUnitRequest, UpdateErpProductUnitRequest, PaginatedKeywordRequest};
use erp_model::response::erp_product_unit::ErpProductUnitResponse;
use crate::convert::erp_product_unit::{create_request_to_model, update_request_to_model, model_to_response};
use anyhow::{Result, anyhow};
use sea_orm::ActiveValue::Set;
use common::constants::enum_constants::{STATUS_DISABLE, STATUS_ENABLE};
use common::base::page::PaginatedResponse;
use common::context::context::LoginUserContext;
use common::interceptor::orm::active_filter::ActiveFilterEntityTrait;

pub async fn create(db: &DatabaseConnection, login_user: LoginUserContext, request: CreateErpProductUnitRequest) -> Result<i64> {
    let mut erp_product_unit = create_request_to_model(&request);
    erp_product_unit.creator = Set(Some(login_user.id));
    erp_product_unit.updater = Set(Some(login_user.id));
    erp_product_unit.tenant_id = Set(login_user.tenant_id);
    let erp_product_unit = erp_product_unit.insert(db).await?;
    Ok(erp_product_unit.id)
}

pub async fn update(db: &DatabaseConnection, login_user: LoginUserContext, request: UpdateErpProductUnitRequest) -> Result<()> {
    let erp_product_unit = ErpProductUnitEntity::find_by_id(request.id)
        .one(db)
        .await?
        .ok_or_else(|| anyhow!("记录未找到"))?;

    let mut erp_product_unit = update_request_to_model(&request, erp_product_unit);
    erp_product_unit.updater = Set(Some(login_user.id));
    erp_product_unit.update(db).await?;
    Ok(())
}

pub async fn delete(db: &DatabaseConnection, login_user: LoginUserContext, id: i64) -> Result<()> {
    let erp_product_unit = ErpProductUnitActiveModel {
        id: Set(id),
        updater: Set(Some(login_user.id)),
        deleted: Set(true),
        ..Default::default()
    };
    erp_product_unit.update(db).await?;
    Ok(())
}

pub async fn get_by_id(db: &DatabaseConnection, login_user: LoginUserContext, id: i64) -> Result<Option<ErpProductUnitResponse>> {
    let condition = Condition::all()
            .add(Column::Id.eq(id))
            .add(Column::TenantId.eq(login_user.tenant_id));
            
    let erp_product_unit = ErpProductUnitEntity::find_active_with_condition(condition)
        .one(db).await?;
    Ok(erp_product_unit.map(model_to_response))
}

pub async fn get_paginated(db: &DatabaseConnection, login_user: LoginUserContext, params: PaginatedKeywordRequest) -> Result<PaginatedResponse<ErpProductUnitResponse>> {
    let condition = Condition::all().add(Column::TenantId.eq(login_user.tenant_id));let paginator = ErpProductUnitEntity::find_active_with_condition(condition)
        .order_by_desc(Column::UpdateTime)
        .paginate(db, params.base.size);

    let total = paginator.num_items().await?;
    let total_pages = (total + params.base.size - 1) / params.base.size; // 向上取整
    let list = paginator
        .fetch_page(params.base.page - 1) // SeaORM 页码从 0 开始，所以减 1
        .await?
        .into_iter()
        .map(model_to_response)
        .collect();

    Ok(PaginatedResponse {
        list,
        total_pages,
        page: params.base.page,
        size: params.base.size,
        total,
    })
}

pub async fn list(db: &DatabaseConnection, login_user: LoginUserContext) -> Result<Vec<ErpProductUnitResponse>> {
    let condition = Condition::all().add(Column::TenantId.eq(login_user.tenant_id));let list = ErpProductUnitEntity::find_active_with_condition(condition)
        .all(db).await?;
    Ok(list.into_iter().map(model_to_response).collect())
}

pub async fn enable(db: &DatabaseConnection, login_user: LoginUserContext, id: i64) -> Result<()> {
    let erp_product_unit = ErpProductUnitActiveModel {
        id: Set(id),
        updater: Set(Some(login_user.id)),
        status: Set(STATUS_ENABLE),
        ..Default::default()
    };
    erp_product_unit.update(db).await?;
    Ok(())
}

pub async fn disable(db: &DatabaseConnection, login_user: LoginUserContext, id: i64) -> Result<()> {
    let erp_product_unit = ErpProductUnitActiveModel {
        id: Set(id),
        updater: Set(Some(login_user.id)),
        status: Set(STATUS_DISABLE),
        ..Default::default()
    };
    erp_product_unit.update(db).await?;
    Ok(())
}

File: admin/src/api/erp_product_unit.ts
import { PaginatedRequest, PaginatedResponse } from '@/base/page';
import { api } from '@/utils/request';

const apis = {
  create: '/erp_product_unit/create', // 新增
  update: '/erp_product_unit/update', // 修改
  delete: '/erp_product_unit/delete', // 删除
  get: '/erp_product_unit/get', // 单条查询
  list: '/erp_product_unit/list', // 列表查询
  page: '/erp_product_unit/page', // 分页查询
  enable: '/erp_product_unit/enable', // 启用
  disable: '/erp_product_unit/disable', // 禁用
}

export interface ErpProductUnitRequest {
  id: number; // 单位ID
  unit_name: string; // 单位名称
  status: number; // 状态
  sort_order: number; // 排序
  remarks: string; // 备注
  department_code: string; // 部门编码
  department_id: number; // 部门ID
}

export interface ErpProductUnitResponse {
  id: number; // 单位ID
  unit_name: string; // 单位名称
  status: number; // 状态
  sort_order: number; // 排序
  remarks: string; // 备注
  department_code: string; // 部门编码
  department_id: number; // 部门ID
  creator: number; // 创建者ID
  create_time: string; // 创建时间
  updater: number; // 更新者ID
  update_time: string; // 更新时间
}

export interface ErpProductUnitQueryCondition extends PaginatedRequest {

}

export const createErpProductUnit = (erp_product_unit: ErpProductUnitRequest): Promise<number> => {
  return api.post<number>(apis.create, erp_product_unit);
}

export const updateErpProductUnit = (erp_product_unit: ErpProductUnitRequest): Promise<void> => {
  return api.post<void>(apis.update, erp_product_unit);
}

export const deleteErpProductUnit = (id: number): Promise<void> => {
  return api.post<void>(`${apis.delete}/${id}`);
}

export const getErpProductUnit = (id: number): Promise<ErpProductUnitResponse> => {
  return api.get<ErpProductUnitResponse>(`${apis.get}/${id}`);
}

export const listErpProductUnit = (): Promise<Array<ErpProductUnitResponse>> => {
  return api.get<Array<ErpProductUnitResponse>>(apis.list);
}

export const pageErpProductUnit = (condition: ErpProductUnitQueryCondition): Promise<PaginatedResponse<ErpProductUnitResponse>> => {
  return api.get<PaginatedResponse<ErpProductUnitResponse>>(apis.page, condition);
}

export const enableErpProductUnit = (id: number): Promise<void> => {
  return api.post<void>(`${apis.enable}/${id}`);
}

export const disableErpProductUnit = (id: number): Promise<void> => {
  return api.post<void>(`${apis.disable}/${id}`);
}